compiler/frontend/spark/src/main/java/edu/snu/onyx/compiler/frontend/spark/JavaRDD.java

```diff
@@ -0,0 +1,209 @@
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Java RDD.
+ * @param <T> type of the final element.
+ */
+public final class JavaRDD<T> extends org.apache.spark.api.java.JavaRDD<T> {
+  private final SparkContext sparkContext;
+  private final Integer parallelism;
+  private final Stack<LoopVertex> loopVertexStack;
+  private final DAG<IRVertex, IREdge> dag;
+  @Nullable private final IRVertex lastVertex;
+  private final Serializer serializer;
+
+  /**
+   * Static method to create a JavaRDD object.
+   * @param sparkContext spark context containing configurations.
+   * @param parallelism parallelism information.
+   * @param <T> type of the resulting object.
+   * @return the new JavaRDD object.
+   */
+  static <T> JavaRDD<T> of(final SparkContext sparkContext, final Integer parallelism) {
+    return new JavaRDD<>(sparkContext, parallelism,
+        new DAGBuilder<IRVertex, IREdge>().buildWithoutSourceSinkCheck(), null);
+  }
+
+  /**
+   * Constructor.
+   * @param sparkContext spark context containing configurations.
+   * @param parallelism parallelism information.
+   * @param dag the current DAG.
+   * @param lastVertex last vertex added to the builder.
+   */
+  private JavaRDD(final SparkContext sparkContext, final Integer parallelism,
+                  final DAG<IRVertex, IREdge> dag, @Nullable final IRVertex lastVertex) {
+    // TODO #366: resolve while implementing scala RDD.
+    super(null, null);
+
+    this.loopVertexStack = new Stack<>();
+    this.sparkContext = sparkContext;
+    this.parallelism = parallelism;
+    this.dag = dag;
+    this.lastVertex = lastVertex;
+    if (sparkContext.conf().get("spark.serializer", "")
+        .equals("org.apache.spark.serializer.KryoSerializer")) {
+      this.serializer = new KryoSerializer(sparkContext.conf());
+    } else {
+      this.serializer = new JavaSerializer(sparkContext.conf());
+    }
+  }
+
+  /////////////// TRANSFORMATIONS ///////////////
+
+  /**
+   * Set initialized source.
+   * @param initialData initial data.
+   * @return the Java RDD with the initialized source vertex.
+   */
+  JavaRDD<T> setSource(final Iterable<T> initialData) {
+    final DAGBuilder<IRVertex, IREdge> builder = new DAGBuilder<>(dag);
+
+    final IRVertex initializedSourceVertex = new InitializedSourceVertex<>(initialData);
+    initializedSourceVertex.setProperty(ParallelismProperty.of(parallelism));
+    builder.addVertex(initializedSourceVertex, loopVertexStack);
+
+    return new JavaRDD<>(this.sparkContext, this.parallelism,
+        builder.buildWithoutSourceSinkCheck(), initializedSourceVertex);
+  }
+
+  /**
+   * Map transform.
+   * @param func function to apply.
+   * @param <O> output type.
+   * @return the JavaRDD with the DAG.
+   */
+  @Override
+  public <O> JavaRDD<O> map(final Function<T, O> func) {
```

Comment:
More details on my previous suggestion on implementing RDD `actions` only:

When an operation is performed on a RDD, for example `rdd_child = rdd_parent.do_sth()`, rdd_child [remembers](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L1688) its parent, rdd_parent. If the operation is an `action`, Spark recursively executes the parent RDDs, starting from the last RDD.

If a given RDD DAG looks like this: src_rdd -> rdd_1 -> rdd_2 -> dst_rdd, the job starts from dst_rdd. Spark checks if dst_rdd is cached, and if not, tries to execute its parent. This goes on recursively until a cached RDD or a source RDD is hit. 

We can mimic this by traversing the RDDs backwards through the pointers to parents just like Spark, but instead of trying to execute the visited RDDS, we can translate each RDD into a corresponding Onyx transform. You can think of it as building the Onyx IR in a reverse order from the sinks.

I'm not sure how this would work out in the actual implementation though. I'm fine with talking about this later, for example when we try to support more RDD operations and would like to not implement the `transformation` methods.

Reply:
Let's discuss about this later. I think it makes it unnecessarily more complex compared to our current implementation, and we will need to implement transformations nevertheless.